lib/integrations/awsoidc/eks_enroll_clusters.go

```diff
@@ -0,0 +1,472 @@
+	eksClusterInfo, err := clt.DescribeCluster(ctx, &eks.DescribeClusterInput{
+		Name: aws.String(clusterName),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	eksCluster := eksClusterInfo.Cluster
+
+	if eksCluster.Status != eksTypes.ClusterStatusActive {
+		return trace.BadParameter(`can't enroll EKS cluster %q - expected "ACTIVE" state, got %q.`, clusterName, eksCluster.Status)
+	}
+
+	principalArn, err := getAccessEntryPrincipalArn(ctx, clt.GetCallerIdentity)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	wasAdded, err := maybeAddAccessEntry(ctx, clusterName, principalArn, clt)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if wasAdded {
+		// If we added access entry, we'll clean it up when function stops executing.
+		defer func() {
+			if _, err := clt.DeleteAccessEntry(ctx, &eks.DeleteAccessEntryInput{
+				ClusterName:  aws.String(clusterName),
+				PrincipalArn: aws.String(principalArn),
+			}); err != nil {
+				log.WithError(err).Warnf("could not delete access entry for principal %q on cluster %q", principalArn, clusterName)
+			}
+		}()
+	}
+
+	_, err = clt.AssociateAccessPolicy(ctx, &eks.AssociateAccessPolicyInput{
+		AccessScope: &eksTypes.AccessScope{
+			Namespaces: nil,
+			Type:       eksTypes.AccessScopeTypeCluster,
+		},
+		ClusterName:  aws.String(clusterName),
+		PolicyArn:    aws.String(eksClusterAdminPolicy),
+		PrincipalArn: aws.String(principalArn),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	kubeConfig, err := getKubeconfig(ctx, clock.Now(), credsProvider, clusterName, req.Region,
+		aws.ToString(eksCluster.CertificateAuthority.Data), aws.ToString(eksCluster.Endpoint))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if alreadyInstalled, err := clt.CheckAgentAlreadyInstalled(kubeConfig, log); err != nil {
+		return trace.Wrap(err, "could not check if teleport-kube-agent is already installed.")
+	} else if alreadyInstalled {
+		return trace.BadParameter("teleport-kube-agent is already installed on the cluster %q", clusterName)
+	}
+
+	if err := clt.InstallKubeAgent(ctx, eksCluster, proxyAddr, kubeConfig, log, req); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+type IdentityGetter func(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+
+func getAccessEntryPrincipalArn(ctx context.Context, identityGetter IdentityGetter) (string, error) {
+	ident, err := identityGetter(ctx, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+
+	}
+
+	parsedARN, err := arn.Parse(aws.ToString(ident.Arn))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	roleParts := strings.Split(parsedARN.Resource, "/")
+	if len(roleParts) < 2 {
+		return "", trace.BadParameter("can't get role name from arn %q", parsedARN.String())
+	}
```

Comment:
I think you can use this function
https://github.com/gravitational/teleport/blob/b9298a474f84b636266992d74deb69752f297cc6/lib/cloud/aws/identity.go#L121

Reply:
Used it, thanks 👍